test/functional/lua/semver_spec.lua

```diff
@@ -0,0 +1,306 @@
+        v1 = '1.0.0-beta',
+        v2 = '1.0.0-beta.2',
+        want = -1,
+      },
+      {
+        desc = 'v1 < v2 when v2 has same first alphabet identifier but has a higher number identifer',
+        v1 = '1.0.0-beta.2',
+        v2 = '1.0.0-beta.11',
+        want = -1,
+      },
+      {
+        desc = 'v1 < v2 when v2 has higher alphabet precedence',
+        v1 = '1.0.0-beta.11',
+        v2 = '1.0.0-rc.1',
+        want = -1,
+      },
+    }
+    for _, tc in ipairs(testcases) do
+      it(
+        string.format('returns %d if %s (v1 = %s, v2 = %s)', tc.want, tc.desc, tc.v1, tc.v2),
+        function()
+          assert.equals(tc.want, semver.cmp(tc.v1, tc.v2, { strict = true }))
+        end
+      )
+    end
+  end)
+
+  describe('parse()', function()
+    describe('parsing', function()
+      describe('strict = true', function()
+        local testcases = {
+          {
+            desc = 'a version without leading "v"',
+            version = '10.20.123',
+            want = {
+              major = 10,
+              minor = 20,
+              patch = 123,
+              prerelease = nil,
+              build = nil,
+            },
+          },
+          {
+            desc = 'a valid version with a leading "v"',
+            version = 'v1.2.3',
+            want = { major = 1, minor = 2, patch = 3 },
+          },
+          {
+            desc = 'a valid version with leading "v" and whitespace',
+            version = '  v1.2.3',
+            want = { major = 1, minor = 2, patch = 3 },
+          },
+          {
+            desc = 'a valid version with leading "v" and trailing whitespace',
+            version = 'v1.2.3  ',
+            want = { major = 1, minor = 2, patch = 3 },
+          },
+          {
+            desc = 'a version with a prerelease',
+            version = '1.2.3-alpha',
+            want = { major = 1, minor = 2, patch = 3, prerelease = 'alpha' },
+          },
+          {
+            desc = 'a version with a prerelease with additional identifiers',
+            version = '1.2.3-alpha.1',
+            want = { major = 1, minor = 2, patch = 3, prerelease = 'alpha.1' },
+          },
+          {
+            desc = 'a version with a build',
+            version = '1.2.3+build.15',
+            want = { major = 1, minor = 2, patch = 3, build = 'build.15' },
+          },
+          {
+            desc = 'a version with a prerelease and build',
+            version = '1.2.3-rc1+build.15',
+            want = {
+              major = 1,
+              minor = 2,
+              patch = 3,
+              prerelease = 'rc1',
+              build = 'build.15',
+            },
+          },
+        }
+        for _, tc in ipairs(testcases) do
+          it(
+            string.format('returns correct table for %q: version = %q', tc.desc, tc.version),
+            function()
+              assert.same(tc.want, semver.parse(tc.version, { strict = true }))
```

Comment:
use `eq()` unless there is a specific reason that assert.same is needed

Reply:
Sorry I might not understand correctly. I tried changing to `assert.equals` here but the tests fail. This is the error
```
            Passed in:
            (table: 0x7fc046161308) {
              [major] = 10
              [minor] = 20
              [patch] = 123 }
            Expected:
            (table: 0x7fc046160168) {
              [major] = 10
              [minor] = 20
              [patch] = 123 }
```
am I misunderstanding something?

Reply:
```lua
local helpers = require('test.functional.helpers')(after_each)
local eq = helpers.eq
```